src/radio_client.c

```diff
@@ -0,0 +1,1122 @@
+    RadioRequest* req)
+{
+    if (req->serial2) {
+        const guint32 used = req->serial2;
+
+        /* Pick another serial */
+        req->serial2 = radio_client_reserve_serial(self);
+        g_hash_table_insert(self->requests, KEY(req->serial2), req);
+        g_hash_table_insert(self->active, KEY(req->serial2),
+            radio_request_ref(req));
+        if (used != req->serial) {
+            g_hash_table_remove(self->requests, KEY(used));
+            g_hash_table_remove(self->pending, KEY(used));
+            g_hash_table_remove(self->active, KEY(used));
+        }
+        radio_request_update_serial(req, req->serial2);
+    } else {
+        /* First submission */
+        req->serial2 = req->serial;
+    }
+
+    /* Actually submit the transaction */
+    req->tx_id = radio_instance_send_request(self->pub.instance,
+        req->code, req->args, radio_client_tx_complete, NULL, self, req);
+    if (req->tx_id) {
+        req->state = RADIO_REQUEST_STATE_PENDING;
+        g_hash_table_insert(self->pending, KEY(req->serial2),
+            radio_request_ref(req));
+        return TRUE;
+    } else {
+        return FALSE;
+    }
+}
+
+static
+gboolean
+radio_client_can_set_owner(
+    RadioClientObject* self,
+    RadioRequestGroup* group)
+{
+    GHashTableIter it;
+    gpointer value;
+
+    g_hash_table_iter_init(&it, self->pending);
+    while (g_hash_table_iter_next(&it, NULL, &value)) {
+        RadioRequest* req = value;
+
+        if (req->group != group) {
+            /*
+             * There's a pending request not associated with any group
+             * or associated with a different group. The specified group
+             * can't become the owner just yet.
+             */
+            return FALSE;
+        }
+    }
+
+    /*
+     * There are no pending requests, or all pending requests are
+     * associated with the specified group. This group can be the owner.
+     */
+    return TRUE;
+}
+
+static
+void
+radio_client_queue_request(
+    RadioClientObject* self,
+    RadioRequest* req)
+{
+    req->state = RADIO_REQUEST_STATE_QUEUED;
+    if (self->queue_last) {
+        self->queue_last->queue_next = req;
+    } else {
+        self->queue_first = req;
+    }
+    self->queue_last = req;
+}
+
+static
+guint
+radio_client_submit_queued_requests(
+    RadioClientObject* self)
+{
+    guint submitted = 0;
+
+    if (!self->block_req && self->pub.instance->connected) {
+        RadioRequest* prev = NULL;
+        RadioRequest* req = self->queue_first;
+
+        while (req) {
+            RadioRequest* next = req->queue_next;
+
+            if (self->owner == req->group ||
+               (!self->owner && !self->owner_queue)) {
+                /* Remove it from the queue */
+                req->queue_next = NULL;
+                if (prev) {
+                    prev->queue_next = next;
+                } else if (!(self->queue_first = next)) {
+                    self->queue_last = NULL;
+                }
```

Comment:
Should we update `next->queue_prev` here too?

Reply:
There's no `queue_prev` )) There are only `queue_first`, `queue_next` and `queue_last`. It's a singly linked list with a pointer to the last element.

Reply:
Ohh, crap, must've been half asleep last night when I read this. Ignore.